{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(filename=u\"main.log\", format=u'%(filename)s[LINE:%(lineno)d]# %(levelname)-8s [%(asctime)s]  %(message)s', level=logging.DEBUG)\n",
    "\n",
    "logging.debug( u'debug' )\n",
    "logging.info( u'info' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import time\n",
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "import json\n",
    "#import Levenshtein as lv\n",
    "import numpy as np\n",
    "from gensim import corpora, models, similarities\n",
    "import spacy\n",
    "\n",
    "import pandas as pd\n",
    "#for findDOI\n",
    "import feedparser\n",
    "import jellyfish\n",
    "import copy\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rss import rss # список рассылки\n",
    "from findDOI import findDOI # функция поиска doi по названию\n",
    "from ALT import altmetrics # функция поиска альтметрик (нужно понять, какие альтметрики нам нужны)\n",
    "from stoplist import stoplist #стоп-лист\n",
    "from rssParser import Parser # класс, который парсит рсс, в датафрейм, - предлагаю его использовать для наполненния\n",
    "from findJournal import findJournal # функция поиска названия журнала по названию статьи"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# p = Parser(rss)\n",
    "# df = p.main()\n",
    "#df.to_csv('example_table.csv')\n",
    "df = pd.read_csv('example_table.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>article_name</th>\n",
       "      <th>doi</th>\n",
       "      <th>journal</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Advances in thermochemical conversion of woody...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;p&gt;Publication date: July–August 2019&lt;/p&gt;&lt;p&gt;&lt;b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Fate of degraded pollutants in waste gas biofi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;p&gt;Publication date: July–August 2019&lt;/p&gt;&lt;p&gt;&lt;b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Strain engineering for microbial production of...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;p&gt;Publication date: July–August 2019&lt;/p&gt;&lt;p&gt;&lt;b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>The state-of-the-art strategies of protein eng...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;p&gt;Publication date: July–August 2019&lt;/p&gt;&lt;p&gt;&lt;b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Engineering of filamentous fungi for efficient...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;p&gt;Publication date: July–August 2019&lt;/p&gt;&lt;p&gt;&lt;b...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                                       article_name  doi journal  \\\n",
       "0           0  Advances in thermochemical conversion of woody...  NaN     NaN   \n",
       "1           1  Fate of degraded pollutants in waste gas biofi...  NaN     NaN   \n",
       "2           2  Strain engineering for microbial production of...  NaN     NaN   \n",
       "3           3  The state-of-the-art strategies of protein eng...  NaN     NaN   \n",
       "4           4  Engineering of filamentous fungi for efficient...  NaN     NaN   \n",
       "\n",
       "                                             summary  \n",
       "0  <p>Publication date: July–August 2019</p><p><b...  \n",
       "1  <p>Publication date: July–August 2019</p><p><b...  \n",
       "2  <p>Publication date: July–August 2019</p><p><b...  \n",
       "3  <p>Publication date: July–August 2019</p><p><b...  \n",
       "4  <p>Publication date: July–August 2019</p><p><b...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Biotechnology Advances'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# example\n",
    "findJournal(df.loc[1,][\"article_name\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "nlp = spacy.load('en', parser=False, ner=False)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Models (gensim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3.7.3'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import gensim \n",
    "gensim.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_path = '/mnt/c/Users/chuvakin/Desktop/ScientificSearch/main/models_lsi/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# загружаем словарь, векторное пространство, модель Lsi (вектороное пространство)\n",
    "# общий корпус натренированный на pubmed \n",
    "dictionary = corpora.Dictionary.load(new_path+\"pubmed_75000docs.dict\")      #Here and later - the first one is PubMed-based LSI-object, other - CUB-based LSI-objects\n",
    "corpus = corpora.MmCorpus(new_path+\"pubmed_5docs.mm\") \n",
    "lsi = models.LsiModel.load(new_path+\"pubmed_75000.lsi\")\n",
    "\n",
    "# химики куб\n",
    "dictionary_CUB_chem = corpora.Dictionary.load(new_path+\"chem.dict\")\n",
    "lsi_CUB_chem = models.LsiModel.load(new_path+\"chem.lsi\")\n",
    "corpus_CUB_chem = corpora.MmCorpus(new_path+\"chem.mm\")\n",
    "\n",
    "# онко куб\n",
    "dictionary_CUB_onco = corpora.Dictionary.load(new_path+\"cancer.dict\")\n",
    "lsi_CUB_onco = models.LsiModel.load(new_path+\"cancer.lsi\")\n",
    "corpus_CUB_onco = corpora.MmCorpus(new_path+\"cancer.mm\")\n",
    "\n",
    "# аутоимунные куб\n",
    "dictionary_CUB_aiz = corpora.Dictionary.load(new_path+\"aiz.dict\")\n",
    "lsi_CUB_aiz = models.LsiModel.load(new_path+\"aiz.lsi\")\n",
    "corpus_CUB_aiz = corpora.MmCorpus(new_path+\"aiz.mm\")\n",
    "\n",
    "# инфекции куб \n",
    "dictionary_CUB_inf = corpora.Dictionary.load(new_path+\"infect.dict\")\n",
    "lsi_CUB_inf = models.LsiModel.load(new_path+\"infect.lsi\")\n",
    "corpus_CUB_inf = corpora.MmCorpus(new_path+\"infect.mm\")\n",
    "\n",
    "# офтальмология куб\n",
    "dictionary_CUB_eye = corpora.Dictionary.load(new_path+\"eye.dict\")\n",
    "lsi_CUB_eye = models.LsiModel.load(new_path+\"eye.lsi\")\n",
    "corpus_CUB_eye = corpora.MmCorpus(new_path+\"eye.mm\")\n",
    "\n",
    "# гететические куб\n",
    "dictionary_CUB_gene = corpora.Dictionary.load(new_path+\"gene.dict\")\n",
    "lsi_CUB_gene = models.LsiModel.load(new_path+\"gene.lsi\")\n",
    "corpus_CUB_gene = corpora.MmCorpus(new_path+\"gene.mm\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'-catenin': 0,\n",
       " 'ability': 1,\n",
       " 'accumulate': 2,\n",
       " 'action': 3,\n",
       " 'active': 4,\n",
       " 'activity': 5,\n",
       " 'add': 6,\n",
       " 'addition': 7,\n",
       " 'additionally': 8,\n",
       " 'analysis': 9,\n",
       " 'annexin': 10,\n",
       " 'antagonist': 11,\n",
       " 'antibody': 12,\n",
       " 'anticancer': 13,\n",
       " 'antitumor': 14,\n",
       " 'apoptosis': 15,\n",
       " 'apoptotic': 16,\n",
       " 'apply': 17,\n",
       " 'assay': 18,\n",
       " 'assess': 19,\n",
       " 'b': 20,\n",
       " 'base': 21,\n",
       " 'bax': 22,\n",
       " 'bc': 23,\n",
       " 'bca': 24,\n",
       " 'bcl-2': 25,\n",
       " 'beij': 26,\n",
       " 'beijing': 27,\n",
       " 'bind': 28,\n",
       " 'blot': 29,\n",
       " 'blotting': 30,\n",
       " 'buffer': 31,\n",
       " 'cadherin': 32,\n",
       " 'cancer': 33,\n",
       " 'cancerous': 34,\n",
       " 'carcinoma': 35,\n",
       " 'cck8': 36,\n",
       " 'cellular': 37,\n",
       " 'centrifuge': 38,\n",
       " 'chamber': 39,\n",
       " 'china': 40,\n",
       " 'clinical': 41,\n",
       " 'colonic': 42,\n",
       " 'colorectal': 43,\n",
       " 'commonly': 44,\n",
       " 'compare': 45,\n",
       " 'concentration': 46,\n",
       " 'considered': 47,\n",
       " 'control': 48,\n",
       " 'culture': 49,\n",
       " 'cyclin': 50,\n",
       " 'd1': 51,\n",
       " 'datum': 52,\n",
       " 'death': 53,\n",
       " 'decrease': 54,\n",
       " 'decreased': 55,\n",
       " 'demonstrate': 56,\n",
       " 'density': 57,\n",
       " 'detect': 58,\n",
       " 'detection': 59,\n",
       " 'development': 60,\n",
       " 'differentiation': 61,\n",
       " 'dilution': 62,\n",
       " 'dmso': 63,\n",
       " 'downregulat': 64,\n",
       " 'due': 65,\n",
       " 'ecl': 66,\n",
       " 'effect': 67,\n",
       " 'effective': 68,\n",
       " 'epithelial': 69,\n",
       " 'establish': 70,\n",
       " 'evaluate': 71,\n",
       " 'evidence': 72,\n",
       " 'exert': 73,\n",
       " 'exhibit': 74,\n",
       " 'exhibited': 75,\n",
       " 'experiment': 76,\n",
       " 'express': 77,\n",
       " 'expression': 78,\n",
       " 'factor': 79,\n",
       " 'fbs': 80,\n",
       " 'finding': 81,\n",
       " 'fitc': 82,\n",
       " 'flow': 83,\n",
       " 'gapdh': 84,\n",
       " 'gradient': 85,\n",
       " 'group': 86,\n",
       " 'growth': 87,\n",
       " 'hct116': 88,\n",
       " 'human': 89,\n",
       " 'identify': 90,\n",
       " 'image': 91,\n",
       " 'imply': 92,\n",
       " 'include': 93,\n",
       " 'increase': 94,\n",
       " 'increased': 95,\n",
       " 'incubate': 96,\n",
       " 'incubation': 97,\n",
       " 'induce': 98,\n",
       " 'induction': 99,\n",
       " 'inhibit': 100,\n",
       " 'inhibited': 101,\n",
       " 'inhibition': 102,\n",
       " 'inhibitory': 103,\n",
       " 'instruction': 104,\n",
       " 'invasion': 105,\n",
       " 'invasive': 106,\n",
       " 'investigate': 107,\n",
       " 'investigation': 108,\n",
       " 'involve': 109,\n",
       " 'kit': 110,\n",
       " 'line': 111,\n",
       " 'lysis': 112,\n",
       " 'ma': 113,\n",
       " 'manufacturer': 114,\n",
       " 'matrigel': 115,\n",
       " 'measure': 116,\n",
       " 'meat': 117,\n",
       " 'mechanism': 118,\n",
       " 'medium': 119,\n",
       " 'member': 120,\n",
       " 'membrane': 121,\n",
       " 'metastasis': 122,\n",
       " 'migrate': 123,\n",
       " 'migration': 124,\n",
       " 'min': 125,\n",
       " 'ml': 126,\n",
       " 'molecular': 127,\n",
       " 'nc': 128,\n",
       " 'negative': 129,\n",
       " 'nk1': 130,\n",
       " 'normal': 131,\n",
       " 'number': 132,\n",
       " 'observe': 133,\n",
       " 'obtain': 134,\n",
       " 'occurrence': 135,\n",
       " 'partly': 136,\n",
       " 'pathway': 137,\n",
       " 'perform': 138,\n",
       " 'pharmacological': 139,\n",
       " 'pi': 140,\n",
       " 'pivotal': 141,\n",
       " 'play': 142,\n",
       " 'possibly': 143,\n",
       " 'potential': 144,\n",
       " 'present': 145,\n",
       " 'progression': 146,\n",
       " 'proliferation': 147,\n",
       " 'promote': 148,\n",
       " 'protein': 149,\n",
       " 'proteintech': 150,\n",
       " 'provide': 151,\n",
       " 'rate': 152,\n",
       " 'receptor': 153,\n",
       " 'reduce': 154,\n",
       " 'reduced': 155,\n",
       " 'regulate': 156,\n",
       " 'regulation': 157,\n",
       " 'relative': 158,\n",
       " 'report': 159,\n",
       " 'result': 160,\n",
       " 'reveal': 161,\n",
       " 'ripa': 162,\n",
       " 'role': 163,\n",
       " 'rpm': 164,\n",
       " 'rpmi-1640': 165,\n",
       " 'secondary': 166,\n",
       " 'series': 167,\n",
       " 'show': 168,\n",
       " 'signal': 169,\n",
       " 'signaling': 170,\n",
       " 'significance': 171,\n",
       " 'significantly': 172,\n",
       " 'sp': 173,\n",
       " 'sps': 174,\n",
       " 'spss': 175,\n",
       " 'stain': 176,\n",
       " 'statistical': 177,\n",
       " 'subsequently': 178,\n",
       " 'suggest': 179,\n",
       " 'supply': 180,\n",
       " 'suppress': 181,\n",
       " 'system': 182,\n",
       " 't': 183,\n",
       " 'target': 184,\n",
       " 'technology': 185,\n",
       " 'therapeutic': 186,\n",
       " 'transwell': 187,\n",
       " 'treat': 188,\n",
       " 'tumor': 189,\n",
       " 'usa': 190,\n",
       " 'vegf': 191,\n",
       " 'viability': 192,\n",
       " 'vivo': 193,\n",
       " 'western': 194,\n",
       " 'wnt': 195,\n",
       " 'work': 196,\n",
       " 'wuhan': 197,\n",
       " 'adult': 198,\n",
       " 'analyse': 199,\n",
       " 'analyze': 200,\n",
       " 'app': 201,\n",
       " 'bioconductor': 202,\n",
       " 'bioinformatic': 203,\n",
       " 'biological': 204,\n",
       " 'blue': 205,\n",
       " 'bone': 206,\n",
       " 'case': 207,\n",
       " 'ch': 208,\n",
       " 'child': 209,\n",
       " 'childhood': 210,\n",
       " 'closely': 211,\n",
       " 'code': 212,\n",
       " 'colon': 213,\n",
       " 'complex': 214,\n",
       " 'conduct': 215,\n",
       " 'construct': 216,\n",
       " 'critical': 217,\n",
       " 'current': 218,\n",
       " 'cycle': 219,\n",
       " 'cytoscape': 220,\n",
       " 'database': 221,\n",
       " 'dataset': 222,\n",
       " 'david': 223,\n",
       " 'deg': 224,\n",
       " 'degree': 225,\n",
       " 'diagnosis': 226,\n",
       " 'diagnostic': 227,\n",
       " 'diagram': 228,\n",
       " 'differentially': 229,\n",
       " 'dna': 230,\n",
       " 'document': 231,\n",
       " 'download': 232,\n",
       " 'downregulated': 233,\n",
       " 'e2f': 234,\n",
       " 'encode': 235,\n",
       " 'enrich': 236,\n",
       " 'enriched': 237,\n",
       " 'enrichment': 238,\n",
       " 'examine': 239,\n",
       " 'family': 240,\n",
       " 'file': 241,\n",
       " 'find': 242,\n",
       " 'fork': 243,\n",
       " 'foxm1': 244,\n",
       " 'frequent': 245,\n",
       " 'functional': 246,\n",
       " 'gene': 247,\n",
       " 'genetic': 248,\n",
       " 'genome': 249,\n",
       " 'geo': 250,\n",
       " 'glycosylation': 251,\n",
       " 'gov': 252,\n",
       " 'green': 253,\n",
       " 'high': 254,\n",
       " 'html': 255,\n",
       " 'http': 256,\n",
       " 'hub': 257,\n",
       " 'important': 258,\n",
       " 'interact': 259,\n",
       " 'involved': 260,\n",
       " 'kinase': 261,\n",
       " 'leukemia': 262,\n",
       " 'li': 263,\n",
       " 'limma': 264,\n",
       " 'liu': 265,\n",
       " 'low': 266,\n",
       " 'lower': 267,\n",
       " 'maintenance': 268,\n",
       " 'malignancy': 269,\n",
       " 'marker': 270,\n",
       " 'marrow': 271,\n",
       " 'match': 272,\n",
       " 'method': 273,\n",
       " 'methylation': 274,\n",
       " 'mirna': 275,\n",
       " 'mitotic': 276,\n",
       " 'module': 277,\n",
       " 'multiple': 278,\n",
       " 'network': 279,\n",
       " 'nod': 280,\n",
       " 'node': 281,\n",
       " 'online': 282,\n",
       " 'org': 283,\n",
       " 'p53': 284,\n",
       " 'package': 285,\n",
       " 'pair': 286,\n",
       " 'plug': 287,\n",
       " 'polymerase': 288,\n",
       " 'poor': 289,\n",
       " 'ppi': 290,\n",
       " 'predict': 291,\n",
       " 'previous': 292,\n",
       " 'probe': 293,\n",
       " 'process': 294,\n",
       " 'prognosis': 295,\n",
       " 'promoter': 296,\n",
       " 'reactome': 297,\n",
       " 'red': 298,\n",
       " 'regulated': 299,\n",
       " 'regulatory': 300,\n",
       " 'relapse': 301,\n",
       " 'relate': 302,\n",
       " 'related': 303,\n",
       " 'relationship': 304,\n",
       " 'remain': 305,\n",
       " 'repair': 306,\n",
       " 'replication': 307,\n",
       " 'represent': 308,\n",
       " 'resistance': 309,\n",
       " 'review': 310,\n",
       " 'sample': 311,\n",
       " 'score': 312,\n",
       " 'screen': 313,\n",
       " 'select': 314,\n",
       " 'sequencing': 315,\n",
       " 'significant': 316,\n",
       " 'specifically': 317,\n",
       " 'stem': 318,\n",
       " 'string': 319,\n",
       " 'strongly': 320,\n",
       " 'survival': 321,\n",
       " 'table': 322,\n",
       " 'tabulate': 323,\n",
       " 'term': 324,\n",
       " 'tfs': 325,\n",
       " 'threshold': 326,\n",
       " 'tool': 327,\n",
       " 'top': 328,\n",
       " 'transcription': 329,\n",
       " 'triangle': 330,\n",
       " 'upregulat': 331,\n",
       " 'upregulate': 332,\n",
       " 'upregulated': 333,\n",
       " 'upregulation': 334,\n",
       " 'variation': 335,\n",
       " 'venn': 336,\n",
       " 'version': 337,\n",
       " 'visualize': 338,\n",
       " 'write': 339,\n",
       " 'www': 340,\n",
       " 'absolute': 341,\n",
       " 'adjuvant': 342,\n",
       " 'administration': 343,\n",
       " 'advance': 344,\n",
       " 'advanced': 345,\n",
       " 'adverse': 346,\n",
       " 'anthracycline': 347,\n",
       " 'anti': 348,\n",
       " 'arm': 349,\n",
       " 'article': 350,\n",
       " 'assign': 351,\n",
       " 'benefit': 352,\n",
       " 'bias': 353,\n",
       " 'blind': 354,\n",
       " 'breast': 355,\n",
       " 'characteristic': 356,\n",
       " 'chemotherapy': 357,\n",
       " 'chen': 358,\n",
       " 'chinese': 359,\n",
       " 'choice': 360,\n",
       " 'ci': 361,\n",
       " 'cochrane': 362,\n",
       " 'common': 363,\n",
       " 'comparable': 364,\n",
       " 'complete': 365,\n",
       " 'conference': 366,\n",
       " 'consensus': 367,\n",
       " 'conservation': 368,\n",
       " 'contrast': 369,\n",
       " 'criterion': 370,\n",
       " 'cyclophosphamide': 371,\n",
       " 'data': 372,\n",
       " 'define': 373,\n",
       " 'difference': 374,\n",
       " 'disease': 375,\n",
       " 'distinct': 376,\n",
       " 'distribution': 377,\n",
       " 'drug': 378,\n",
       " 'dual': 379,\n",
       " 'early': 380,\n",
       " 'eligible': 381,\n",
       " 'endorse': 382,\n",
       " 'epidermal': 383,\n",
       " 'equivalent': 384,\n",
       " 'estimate': 385,\n",
       " 'event': 386,\n",
       " 'examination': 387,\n",
       " 'extract': 388,\n",
       " 'fluorouracil': 389,\n",
       " 'follow': 390,\n",
       " 'forest': 391,\n",
       " 'free': 392,\n",
       " 'full': 393,\n",
       " 'funnel': 394,\n",
       " 'future': 395,\n",
       " 'gain': 396,\n",
       " 'hazard': 397,\n",
       " 'her2': 398,\n",
       " 'her2-positive': 399,\n",
       " 'heterogeneity': 400,\n",
       " 'hormone': 401,\n",
       " 'hr': 402,\n",
       " 'hrs': 403,\n",
       " 'i2': 404,\n",
       " 'identification': 405,\n",
       " 'improve': 406,\n",
       " 'initial': 407,\n",
       " 'investigator': 408,\n",
       " 'item': 409,\n",
       " 'jin': 410,\n",
       " 'lapatinib': 411,\n",
       " 'large': 412,\n",
       " 'literature': 413,\n",
       " 'locally': 414,\n",
       " 'locoregional': 415,\n",
       " 'log': 416,\n",
       " 'long': 417,\n",
       " 'main': 418,\n",
       " 'management': 419,\n",
       " 'mastectomy': 420,\n",
       " 'meta': 421,\n",
       " 'model': 422,\n",
       " 'nac': 423,\n",
       " 'neoadjuvant': 424,\n",
       " 'observed': 425,\n",
       " 'oncology': 426,\n",
       " 'operable': 427,\n",
       " 'optimum': 428,\n",
       " 'os': 429,\n",
       " 'outcome': 430,\n",
       " 'panel': 431,\n",
       " 'past': 432,\n",
       " 'pathological': 433,\n",
       " 'patient': 434,\n",
       " 'pcr': 435,\n",
       " 'period': 436,\n",
       " 'plot': 437,\n",
       " 'pool': 438,\n",
       " 'postoperative': 439,\n",
       " 'predictive': 440,\n",
       " 'prefer': 441,\n",
       " 'preferred': 442,\n",
       " 'previously': 443,\n",
       " 'primary': 444,\n",
       " 'publication': 445,\n",
       " 'publish': 446,\n",
       " 'radiological': 447,\n",
       " 'radiotherapy': 448,\n",
       " 'randomize': 449,\n",
       " 'ratio': 450,\n",
       " 'rct': 451,\n",
       " 'receive': 452,\n",
       " 'recent': 453,\n",
       " 'recently': 454,\n",
       " 'recurrence': 455,\n",
       " 'regimen': 456,\n",
       " 'relevant': 457,\n",
       " 'reporting': 458,\n",
       " 'residual': 459,\n",
       " 'resistant': 460,\n",
       " 'response': 461,\n",
       " 'restrict': 462,\n",
       " 'rfs': 463,\n",
       " 'risk': 464,\n",
       " 'search': 465,\n",
       " 'selection': 466,\n",
       " 'shi': 467,\n",
       " 'short': 468,\n",
       " 'similar': 469,\n",
       " 'size': 470,\n",
       " 'society': 471,\n",
       " 'software': 472,\n",
       " 'stage': 473,\n",
       " 'statement': 474,\n",
       " 'status': 475,\n",
       " 'subgroup': 476,\n",
       " 'subset': 477,\n",
       " 'subtype': 478,\n",
       " 'superior': 479,\n",
       " 'surgery': 480,\n",
       " 'symmetrical': 481,\n",
       " 'systematic': 482,\n",
       " 'systemic': 483,\n",
       " 'tamoxifen': 484,\n",
       " 'taxane': 485,\n",
       " 'text': 486,\n",
       " 'therapy': 487,\n",
       " 'time': 488,\n",
       " 'translate': 489,\n",
       " 'trastuzumab': 490,\n",
       " 'trial': 491,\n",
       " 'triple': 492,\n",
       " 'type': 493,\n",
       " 'unclear': 494,\n",
       " 'update': 495,\n",
       " 'wang': 496,\n",
       " 'woman': 497,\n",
       " 'yan': 498,\n",
       " 'yang': 499,\n",
       " 'year': 500,\n",
       " 'achieve': 501,\n",
       " 'adenocarcinoma': 502,\n",
       " 'adequate': 503,\n",
       " 'age': 504,\n",
       " 'approve': 505,\n",
       " 'await': 506,\n",
       " 'baseline': 507,\n",
       " 'calculate': 508,\n",
       " 'calculated': 509,\n",
       " 'capecitabine': 510,\n",
       " 'cardiac': 511,\n",
       " 'caution': 512,\n",
       " 'chemotherapeutic': 513,\n",
       " 'cisplatin': 514,\n",
       " 'combination': 515,\n",
       " 'combine': 516,\n",
       " 'completion': 517,\n",
       " 'conclusion': 518,\n",
       " 'consent': 519,\n",
       " 'count': 520,\n",
       " 'date': 521,\n",
       " 'day': 522,\n",
       " 'delay': 523,\n",
       " 'diarrhea': 524,\n",
       " 'docetaxel': 525,\n",
       " 'dos': 526,\n",
       " 'dose': 527,\n",
       " 'efficacy': 528,\n",
       " 'ejection': 529,\n",
       " 'enrol': 530,\n",
       " 'exclude': 531,\n",
       " 'feasibility': 532,\n",
       " 'finally': 533,\n",
       " 'fish': 534,\n",
       " 'fluoropyrimidine': 535,\n",
       " 'fraction': 536,\n",
       " 'fu': 537,\n",
       " 'gastric': 538,\n",
       " 'gastroesophageal': 539,\n",
       " 'grade': 540,\n",
       " 'humanize': 541,\n",
       " 'ihc': 542,\n",
       " 'iii': 543,\n",
       " 'infusion': 544,\n",
       " 'kaplan': 545,\n",
       " 'kg': 546,\n",
       " 'leave': 547,\n",
       " 'limitation': 548,\n",
       " 'logic': 549,\n",
       " 'm2': 550,\n",
       " 'median': 551,\n",
       " 'metastatic': 552,\n",
       " 'methodology': 553,\n",
       " 'mm3': 554,\n",
       " 'month': 555,\n",
       " 'multicenter': 556,\n",
       " 'oxaliplatin': 557,\n",
       " 'partial': 558,\n",
       " 'performance': 559,\n",
       " 'pertuzumab': 560,\n",
       " 'pfs': 561,\n",
       " 'phase': 562,\n",
       " 'positive': 563,\n",
       " 'preliminary': 564,\n",
       " 'progressive': 565,\n",
       " 'randomized': 566,\n",
       " 'range': 567,\n",
       " 'require': 568,\n",
       " 'safety': 569,\n",
       " 'severe': 570,\n",
       " 'site': 571,\n",
       " 'small': 572,\n",
       " 'stable': 573,\n",
       " 'standard': 574,\n",
       " 'stomatitis': 575,\n",
       " 'subsequent': 576,\n",
       " 'supplementary': 577,\n",
       " 'test': 578,\n",
       " 'tested': 579,\n",
       " 'toxicity': 580,\n",
       " 'triplet': 581,\n",
       " 'unacceptable': 582,\n",
       " 'untreat': 583,\n",
       " 'untreated': 584,\n",
       " 'ventricular': 585,\n",
       " 'abnormal': 586,\n",
       " 'accompany': 587,\n",
       " 'accurate': 588,\n",
       " 'affect': 589,\n",
       " 'affiliate': 590,\n",
       " 'affiliated': 591,\n",
       " 'alveolar': 592,\n",
       " 'area': 593,\n",
       " 'aspect': 594,\n",
       " 'atypical': 595,\n",
       " 'biopsy': 596,\n",
       " 'cavity': 597,\n",
       " 'cd20': 598,\n",
       " 'cervical': 599,\n",
       " 'chop': 600,\n",
       " 'classify': 601,\n",
       " 'compose': 602,\n",
       " 'compute': 603,\n",
       " 'confirm': 604,\n",
       " 'consistent': 605,\n",
       " 'conventional': 606,\n",
       " 'cortical': 607,\n",
       " 'ct': 608,\n",
       " 'dental': 609,\n",
       " 'desirable': 610,\n",
       " 'differential': 611,\n",
       " 'diffuse': 612,\n",
       " 'discomfort': 613,\n",
       " 'dlbcl': 614,\n",
       " 'doctor': 615,\n",
       " 'extraction': 616,\n",
       " 'extranodal': 617,\n",
       " 'favorable': 618,\n",
       " 'final': 619,\n",
       " 'frequency': 620,\n",
       " 'frequently': 621,\n",
       " 'gland': 622,\n",
       " 'histologic': 623,\n",
       " 'history': 624,\n",
       " 'hospital': 625,\n",
       " 'immunohistochemical': 626,\n",
       " 'immunohistochemistry': 627,\n",
       " 'incisional': 628,\n",
       " 'inflammatory': 629,\n",
       " 'inhibitor': 630,\n",
       " 'involvement': 631,\n",
       " 'lesion': 632,\n",
       " 'local': 633,\n",
       " 'loss': 634,\n",
       " 'lymph': 635,\n",
       " 'lymphoid': 636,\n",
       " 'lymphoma': 637,\n",
       " 'magnification': 638,\n",
       " 'majority': 639,\n",
       " 'male': 640,\n",
       " 'malignant': 641,\n",
       " 'manuscript': 642,\n",
       " 'mass': 643,\n",
       " 'maxillary': 644,\n",
       " 'misdiagnos': 645,\n",
       " 'molar': 646,\n",
       " 'mouth': 647,\n",
       " 'mucosa': 648,\n",
       " 'nhl': 649,\n",
       " 'nonspecific': 650,\n",
       " 'numbness': 651,\n",
       " 'occur': 652,\n",
       " 'oral': 653,\n",
       " 'original': 654,\n",
       " 'osteolytic': 655,\n",
       " 'pain': 656,\n",
       " 'pax5': 657,\n",
       " 'periodontal': 658,\n",
       " 'posterior': 659,\n",
       " 'presentation': 660,\n",
       " 'radiation': 661,\n",
       " 'radiography': 662,\n",
       " 'rare': 663,\n",
       " 'region': 664,\n",
       " 'ring': 665,\n",
       " 'rituximab': 666,\n",
       " 'scan': 667,\n",
       " 'sign': 668,\n",
       " 'solid': 669,\n",
       " 'song': 670,\n",
       " 'squamous': 671,\n",
       " 'subdivide': 672,\n",
       " 'swell': 673,\n",
       " 'symptom': 674,\n",
       " 'tomography': 675,\n",
       " 'tooth': 676,\n",
       " 'university': 677,\n",
       " 'upper': 678,\n",
       " 'writing': 679,\n",
       " 'amyloid': 680,\n",
       " 'approximately': 681,\n",
       " 'avoid': 682,\n",
       " 'axonal': 683,\n",
       " 'basal': 684,\n",
       " 'bilateral': 685,\n",
       " 'blood': 686,\n",
       " 'body': 687,\n",
       " 'brain': 688,\n",
       " 'callosum': 689,\n",
       " 'capillary': 690,\n",
       " 'category': 691,\n",
       " 'cbc': 692,\n",
       " 'cd3': 693,\n",
       " 'cd43': 694,\n",
       " 'central': 695,\n",
       " 'cerebral': 696,\n",
       " 'classification': 697,\n",
       " 'clinician': 698,\n",
       " 'cns': 699,\n",
       " 'comprehensive': 700,\n",
       " 'condition': 701,\n",
       " 'cord': 702,\n",
       " 'corpus': 703,\n",
       " 'dai': 704,\n",
       " 'deoxynucleotidyl': 705,\n",
       " 'dependent': 706,\n",
       " 'draft': 707,\n",
       " 'encounter': 708,\n",
       " 'enhancement': 709,\n",
       " 'enlargement': 710,\n",
       " 'external': 711,\n",
       " 'failure': 712,\n",
       " 'feature': 713,\n",
       " 'form': 714,\n",
       " 'frontal': 715,\n",
       " 'ganglion': 716,\n",
       " 'general': 717,\n",
       " 'gyrus': 718,\n",
       " 'head': 719,\n",
       " 'histological': 720,\n",
       " 'illness': 721,\n",
       " 'imaging': 722,\n",
       " 'infiltrate': 723,\n",
       " 'information': 724,\n",
       " 'injury': 725,\n",
       " 'interstitium': 726,\n",
       " 'intracranial': 727,\n",
       " 'ion': 728,\n",
       " 'limb': 729,\n",
       " 'liver': 730,\n",
       " 'lymphoblast': 731,\n",
       " 'lymphoblastic': 732,\n",
       " 'lymphocyte': 733,\n",
       " 'magnetic': 734,\n",
       " 'man': 735,\n",
       " 'manifestation': 736,\n",
       " 'medical': 737,\n",
       " 'medulla': 738,\n",
       " 'midbrain': 739,\n",
       " 'misdiagnosis': 740,\n",
       " 'mpo': 741,\n",
       " 'mri': 742,\n",
       " 'neoplasm': 743,\n",
       " 'neoplastic': 744,\n",
       " 'nerve': 745,\n",
       " 'nervous': 746,\n",
       " 'optic': 747,\n",
       " 'optical': 748,\n",
       " 'organ': 749,\n",
       " 'originate': 750,\n",
       " 'peripheral': 751,\n",
       " 'perivascular': 752,\n",
       " 'phenotype': 753,\n",
       " 'physical': 754,\n",
       " 'possibility': 755,\n",
       " 'post': 756,\n",
       " 'precursor': 757,\n",
       " 'prior': 758,\n",
       " 'pronounce': 759,\n",
       " 'prospect': 760,\n",
       " 'record': 761,\n",
       " 'reflect': 762,\n",
       " 'resonance': 763,\n",
       " 'scatter': 764,\n",
       " 'smear': 765,\n",
       " 'space': 766,\n",
       " 'spleen': 767,\n",
       " 'subarachnoid': 768,\n",
       " 't2': 769,\n",
       " 'tdt': 770,\n",
       " 'terminal': 771,\n",
       " 'tissue': 772,\n",
       " 'tonsil': 773,\n",
       " 'transferase': 774,\n",
       " 'trauma': 775,\n",
       " 'understand': 776,\n",
       " 'understanding': 777,\n",
       " 'valuable': 778,\n",
       " 'vessel': 779,\n",
       " 'wbc': 780,\n",
       " 'wbcs': 781,\n",
       " 'wu': 782,\n",
       " 'abdominal': 783,\n",
       " 'adenomatous': 784,\n",
       " 'adhere': 785,\n",
       " 'adjacent': 786,\n",
       " 'aggressive': 787,\n",
       " 'anastomotic': 788,\n",
       " 'animal': 789,\n",
       " 'appearance': 790,\n",
       " 'arrow': 791,\n",
       " 'artery': 792,\n",
       " 'assessment': 793,\n",
       " 'benign': 794,\n",
       " 'bland': 795,\n",
       " 'bleed': 796,\n",
       " 'bleeding': 797,\n",
       " 'bloc': 798,\n",
       " 'boundary': 799,\n",
       " 'bowel': 800,\n",
       " 'branch': 801,\n",
       " 'choose': 802,\n",
       " 'cm': 803,\n",
       " 'cold': 804,\n",
       " 'comorbidity': 805,\n",
       " 'complicated': 806,\n",
       " 'contraindicate': 807,\n",
       " 'curative': 808,\n",
       " 'damage': 809,\n",
       " 'dense': 810,\n",
       " 'digestive': 811,\n",
       " 'distal': 812,\n",
       " 'duodenum': 813,\n",
       " 'enlarge': 814,\n",
       " 'excellent': 815,\n",
       " 'excision': 816,\n",
       " 'experience': 817,\n",
       " 'extremely': 818,\n",
       " 'fap': 819,\n",
       " 'field': 820,\n",
       " 'flush': 821,\n",
       " 'germline': 822,\n",
       " 'infection': 823,\n",
       " 'intestinal': 824,\n",
       " 'intestine': 825,\n",
       " 'isolate': 826,\n",
       " 'jejunum': 827,\n",
       " 'jun': 828,\n",
       " 'lu': 829,\n",
       " 'margin': 830,\n",
       " 'mesenteric': 831,\n",
       " 'motivated': 832,\n",
       " 'mutation': 833,\n",
       " 'needle': 834,\n",
       " 'nuclear': 835,\n",
       " 'nucleus': 836,\n",
       " 'nutrition': 837,\n",
       " 'obstacle': 838,\n",
       " 'operation': 839,\n",
       " 'operative': 840,\n",
       " 'part': 841,\n",
       " 'pathologic': 842,\n",
       " 'polyp': 843,\n",
       " 'polyposis': 844,\n",
       " 'portion': 845,\n",
       " 'preoperative': 846,\n",
       " 'preservation': 847,\n",
       " 'procedure': 848,\n",
       " 'protect': 849,\n",
       " 'proximal': 850,\n",
       " 'recovery': 851,\n",
       " 'remove': 852,\n",
       " 'research': 853,\n",
       " 'resection': 854,\n",
       " 'root': 855,\n",
       " 'section': 856,\n",
       " 'sma': 857,\n",
       " 'smv': 858,\n",
       " 'soft': 859,\n",
       " 'solution': 860,\n",
       " 'specific': 861,\n",
       " 'speciman': 862,\n",
       " 'spindle': 863,\n",
       " 'staining': 864,\n",
       " 'step': 865,\n",
       " 'stoma': 866,\n",
       " 'suitable': 867,\n",
       " 'surface': 868,\n",
       " 'surgeon': 869,\n",
       " 'surgical': 870,\n",
       " 'surround': 871,\n",
       " 'syndrome': 872,\n",
       " 'technique': 873,\n",
       " 'tightly': 874,\n",
       " 'tract': 875,\n",
       " 'transverse': 876,\n",
       " 'ultrasonography': 877,\n",
       " 'undergo': 878,\n",
       " 'underwent': 879,\n",
       " 'union': 880,\n",
       " 'vascular': 881,\n",
       " 'vein': 882,\n",
       " 'visualization': 883,\n",
       " 'wall': 884,\n",
       " 'wisconsin': 885,\n",
       " 'xiao': 886,\n",
       " 'abdoman': 887,\n",
       " 'abdomen': 888,\n",
       " 'acc': 889,\n",
       " 'adrenal': 890,\n",
       " 'afp': 891,\n",
       " 'alkaline': 892,\n",
       " 'alpha': 893,\n",
       " 'arise': 894,\n",
       " 'cd30': 895,\n",
       " 'cga': 896,\n",
       " 'clear': 897,\n",
       " 'cluster': 898,\n",
       " 'codon': 899,\n",
       " 'cytoplasm': 900,\n",
       " 'd2': 901,\n",
       " 'develop': 902,\n",
       " 'ema': 903,\n",
       " 'english': 904,\n",
       " 'eosinophilic': 905,\n",
       " 'epithelioid': 906,\n",
       " 'familial': 907,\n",
       " 'fetoprotein': 908,\n",
       " 'fibrous': 909,\n",
       " 'fibrovascular': 910,\n",
       " 'germ': 911,\n",
       " 'lymphocyt': 912,\n",
       " 'macroscopically': 913,\n",
       " 'mantle': 914,\n",
       " 'microscopically': 915,\n",
       " 'mismatch': 916,\n",
       " 'mixed': 917,\n",
       " 'mlh1': 918,\n",
       " 'msh2': 919,\n",
       " 'msh6': 920,\n",
       " 'necrosis': 921,\n",
       " 'nest': 922,\n",
       " 'nucleoli': 923,\n",
       " 'osteosarcoma': 924,\n",
       " 'ovarian': 925,\n",
       " 'ovary': 926,\n",
       " 'papillary': 927,\n",
       " 'placental': 928,\n",
       " 'pms2': 929,\n",
       " 'polygonal': 930,\n",
       " 'positivity': 931,\n",
       " 'primitive': 932,\n",
       " 'reticular': 933,\n",
       " 'sac': 934,\n",
       " 'separate': 935,\n",
       " 'septa': 936,\n",
       " 'serum': 937,\n",
       " 'sheet': 938,\n",
       " 'sporadic': 939,\n",
       " 'stroma': 940,\n",
       " 'structure': 941,\n",
       " 'syn': 942,\n",
       " 'synchronous': 943,\n",
       " 'tubular': 944,\n",
       " 'yolk': 945,\n",
       " 'advantage': 946,\n",
       " 'allocation': 947,\n",
       " 'anastomosis': 948,\n",
       " 'attempt': 949,\n",
       " 'bayesian': 950,\n",
       " 'bile': 951,\n",
       " 'cai': 952,\n",
       " 'carcinogenesis': 953,\n",
       " 'close': 954,\n",
       " 'comparative': 955,\n",
       " 'comparison': 956,\n",
       " 'complication': 957,\n",
       " 'consistency': 958,\n",
       " 'controversial': 959,\n",
       " 'corp': 960,\n",
       " 'country': 961,\n",
       " 'cri': 962,\n",
       " 'cumulative': 963,\n",
       " 'curve': 964,\n",
       " 'derive': 965,\n",
       " 'direct': 966,\n",
       " 'disagreement': 967,\n",
       " 'discussion': 968,\n",
       " 'embase': 969,\n",
       " 'empty': 970,\n",
       " 'endoscopic': 971,\n",
       " 'esophagitis': 972,\n",
       " 'exceed': 973,\n",
       " 'food': 974,\n",
       " 'gastrectomy': 975,\n",
       " 'gastritis': 976,\n",
       " 'gc': 977,\n",
       " 'graphical': 978,\n",
       " 'ii': 979,\n",
       " 'inclusion': 980,\n",
       " 'inconsistency': 981,\n",
       " 'independently': 982,\n",
       " 'indirect': 983,\n",
       " 'intervention': 984,\n",
       " 'intraoperative': 985,\n",
       " 'leakage': 986,\n",
       " 'library': 987,\n",
       " 'loop': 988,\n",
       " 'morbidity': 989,\n",
       " 'odd': 990,\n",
       " 'ovid': 991,\n",
       " 'pass': 992,\n",
       " 'passage': 993,\n",
       " 'perioperative': 994,\n",
       " 'physiological': 995,\n",
       " 'prevent': 996,\n",
       " 'pubm': 997,\n",
       " 'pubmed': 998,\n",
       " 'quality': 999,\n",
       " ...}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dictionary.token2id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### indexes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'u'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-8a5df432fe6d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#index = similarities.MatrixSimilarity(lsi[corpus])          #Indexes\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mindex_CUB_chem\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msimilarities\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mMatrixSimilarity\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlsi_CUB_onco\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcorpus_CUB_chem\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mindex_CUB_onco\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msimilarities\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mMatrixSimilarity\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlsi_CUB_onco\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcorpus_CUB_onco\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mindex_CUB_aiz\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msimilarities\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mMatrixSimilarity\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlsi_CUB_aiz\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcorpus_CUB_aiz\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mindex_CUB_inf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msimilarities\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mMatrixSimilarity\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlsi_CUB_aiz\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcorpus_CUB_inf\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/gensim/models/lsimodel.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, bow, scaled, chunksize)\u001b[0m\n\u001b[1;32m    566\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    567\u001b[0m         \"\"\"\n\u001b[0;32m--> 568\u001b[0;31m         \u001b[0;32massert\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprojection\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mu\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"decomposition not initialized yet\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    569\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    570\u001b[0m         \u001b[0;31m# if the input vector is in fact a corpus, return a transformed corpus as a result\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'u'"
     ]
    }
   ],
   "source": [
    "index = similarities.MatrixSimilarity(lsi[corpus])          #Indexes\n",
    "index_CUB_chem = similarities.MatrixSimilarity(lsi_CUB_onco[corpus_CUB_chem])\n",
    "index_CUB_onco = similarities.MatrixSimilarity(lsi_CUB_onco[corpus_CUB_onco])\n",
    "index_CUB_aiz = similarities.MatrixSimilarity(lsi_CUB_aiz[corpus_CUB_aiz])\n",
    "index_CUB_inf = similarities.MatrixSimilarity(lsi_CUB_aiz[corpus_CUB_inf])\n",
    "index_CUB_eye = similarities.MatrixSimilarity(lsi_CUB_aiz[corpus_CUB_eye])\n",
    "index_CUB_gene = similarities.MatrixSimilarity(lsi_CUB_aiz[corpus_CUB_gene])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
